### ISSUE TITLE
[Java] Loading frozen graph causes a segfault

### ISSUE THREAD

**Issue by tianwang95** [tensorflow/tensorflow#7952]
I froze then saved a tensorflow graph in python with [freeze_graph.py](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/tools/freeze_graph.py) producing a single model.pb file.

However, loading the graph in Java (where modelStream is an InputStream from reading the .pb file):
```
Graph g = new Graph();
InputStream modelStream = new FileInputStream("/path/to/model.pb");
g.importGraphDef(IOUtils.toByteArray(modelStream);
```
causes a segfault:
```
# A fatal error has been detected by the Java Runtime Environment:
#
#  SIGSEGV (0xb) at pc=0x0000000125526b1c, pid=6482, tid=7171
#
# JRE version: Java(TM) SE Runtime Environment (8.0_91-b14) (build 1.8.0_91-b14)
# Java VM: Java HotSpot(TM) 64-Bit Server VM (25.91-b14 mixed mode bsd-amd64 compressed oops)
# Problematic frame:
# C  [libtensorflow_jni.dylib+0x1bfdb1c]  tensorflow::shape_inference::InferenceContext::WithRank(tensorflow::shape_inference::ShapeHandle, int, tensorflow::shape_inference::ShapeHandle*)+0x1c
#
```
The same model.pb file loads and runs properly in Python.

### Environment info
Operating System: MacOS Sierra 10.12.3
Native library for OSX downloaded from [here](https://storage.googleapis.com/tensorflow/libtensorflow/libtensorflow_jni-cpu-darwin-x86_64-1.0.0-PREVIEW1.tar.gz) by following the [readme on the Java Tensorflow repo](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/java/README.md)

**Comment by asimshankar** [comment]
Thanks for the report. Can you share instructions to reproduce (if possible, a toy model and the `freeze_graph.py` invocation used)?

It seems this might have something to do with the model or with `freeze_graph`. For example, this doesn't happen with the frozen inception graph used by the [LabelImage example in the README](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/java/README.md), so any more information on reproducing the problem will be helpful.

**Comment by tianwang95** [comment]
The model I used is quite bulky and was originally developed in Keras (2 inputs -> GRU -> concat -> 2 feed forward steps -> sigmoid) so not sure if that will end up being helpful (although I will try and work on a toy model that reproduces the bug). 

However, this is the freeze_graph invocation I used:

```
    checkpoint_name = "saved_checkpoint"
    checkpoint_prefix = os.path.join(temp_dir, checkpoint_name)
    checkpoint_state_name = "checkpoint_state"
    input_graph_name = "input_graph.pb"
    output_graph_name = "model.pb"
    input_graph_path = os.path.join(temp_dir, input_graph_name)
    input_checkpoint_path = os.path.join(temp_dir, checkpoint_name)
    output_graph_path = os.path.join(export_dir, output_graph_name)

    #### Save the model to a checkpoint
    sess = K.get_session()
    saver = saver_lib.Saver()
    checkpoint_path = saver.save(
            sess,
            checkpoint_prefix,
            global_step=0,
            latest_filename=checkpoint_state_name)

    ### Save graph definition
    tf.train.write_graph(sess.graph.as_graph_def(), temp_dir, input_graph_name)

    ### Freeze graph
    output_node_names = "Sigmoid"
    restore_op_name = "save/restore_all"
    filename_tensor_name = "save/Const:0"
    input_saver_def_path = ""
    input_binary = False
    clear_devices = True

    freeze_graph.freeze_graph(input_graph_path,
                              input_saver_def_path,
                              input_binary,
                              checkpoint_path,
                              output_node_names,
                              restore_op_name,
                              filename_tensor_name,
                              output_graph_path,
                              clear_devices,
                              "")
```
I am exporting the graph from Keras, so it is possible that its an issue on their side, although the model does load fine afterwards with:
```
#### Test to make sure it has worked
    graph = load_graph(output_graph_path)
    q1_input = graph.get_tensor_by_name('import/q1_input:0')
    q2_input = graph.get_tensor_by_name('import/q2_input:0')
    y_tensor = graph.get_tensor_by_name('import/Sigmoid:0')

    with tf.Session(graph = graph) as session:
        acc = 0.0;
        for point in data.dev_generator():
            x1 = point[0][0]
            x2 = point[0][1]
            y = point[1]
            y_out = session.run(y_tensor, feed_dict = {
                    q1_input: x1,
                    q2_input: x2,
                })
            acc += np.sum(y == np.round(y_out)) / 64.0
        print("Final acc: {}".format(acc / num))
```

**Comment by tatatodd** [comment]
@tianwang95 an example model that causes the crash would be great, thanks!

@asimshankar Before we have that, do you have any other ideas?

**Comment by prb12** [comment]
Automatically closing due to lack of recent activity. Please reopen when further information becomes available. Thanks!

**Comment by dadib** [comment]
I am running into the same issue and have reduced it to a minimal failing example.
```shell
dbSpotify:~ dadi$ sw_vers
ProductName:	Mac OS X
ProductVersion:	10.12
BuildVersion:	16A323
```

The problem seems to be in dynamic_rnn. When we unroll the network manually rather than use dynamic_rnn it works fine.
```python
import tensorflow as tf


def define_graph():
    features = tf.placeholder(tf.float32, [1, None, 3], name='input')
    cell = tf.contrib.rnn.BasicLSTMCell(20)
    output, _ = tf.nn.dynamic_rnn(cell, features, dtype=tf.float32)
    tf.identity(output, name='output')

if __name__ == '__main__':
    print("Tensorflow version: " + tf.__version__)
    define_graph()
    with tf.Session() as sess:
        sess.run([tf.global_variables_initializer()])
        in_graph_def = tf.get_default_graph().as_graph_def()
        out_graph_def = tf.graph_util.convert_variables_to_constants(
            sess, in_graph_def, ['output'])

    with tf.gfile.GFile('frozen_model.pb', 'wb') as f:
        f.write(out_graph_def.SerializeToString())

```

```shell
dbSpotify:models dadi$ python example.py
Tensorflow version: 1.0.1
W tensorflow/core/platform/cpu_feature_guard.cc:45] The TensorFlow library wasn't compiled to use SSE4.1 instructions, but these are available on your machine and could speed up CPU computations.
W tensorflow/core/platform/cpu_feature_guard.cc:45] The TensorFlow library wasn't compiled to use SSE4.2 instructions, but these are available on your machine and could speed up CPU computations.
W tensorflow/core/platform/cpu_feature_guard.cc:45] The TensorFlow library wasn't compiled to use AVX instructions, but these are available on your machine and could speed up CPU computations.
W tensorflow/core/platform/cpu_feature_guard.cc:45] The TensorFlow library wasn't compiled to use AVX2 instructions, but these are available on your machine and could speed up CPU computations.
W tensorflow/core/platform/cpu_feature_guard.cc:45] The TensorFlow library wasn't compiled to use FMA instructions, but these are available on your machine and could speed up CPU computations.
Converted 2 variables to const ops.
```
Dowloaded jar / CPU-only native libs as per [instructions](https://github.com/tensorflow/tensorflow/tree/master/tensorflow/java)
Loading the model in java:
```java
package com.spotify;

import java.io.File;
import java.io.IOException;
import java.nio.file.Files;
import org.junit.Test;
import org.tensorflow.Graph;

public class TfModelLoadTest {

  @Test
  public void testImportTfModel() throws IOException {
    final File modelFile = new File("./frozen_model.pb");
    byte[] graphDef = Files.readAllBytes(modelFile.toPath());
    Graph graph = new Graph();
    graph.importGraphDef(graphDef);
  }

}

```

Leads to this [error report](https://gist.github.com/dadib/6270b70797997b0b0d9163cf4b8877ce)

**Comment by asimshankar** [comment]
@dadib: Thanks for providing a minimal example. 

Digging a bit, I believe the same issues as in #8284 that should be fixed at head (and will be fixed in the next release - 1.1). Could you try with the nightly builds:

- Java source: https://ci.tensorflow.org/view/Nightly/job/nightly-libtensorflow/TYPE=mac-slave/lastSuccessfulBuild/artifact/lib_package/libtensorflow.jar
- JNI: https://ci.tensorflow.org/view/Nightly/job/nightly-libtensorflow/TYPE=mac-slave/lastSuccessfulBuild/artifact/lib_package/libtensorflow_jni-cpu-darwin-x86_64.tar.gz

And then switch to 1.1 when that is released?

**Comment by dadib** [comment]
@asimshankar works with those nightly builds. Thanks!
